File: app/src/main/java/ru/orangesoftware/financisto/rates/FreeCurrencyRateDownloader.java
package ru.orangesoftware.financisto.rates;

import org.json.JSONObject;

import ru.orangesoftware.financisto.app.DependenciesHolder;
import ru.orangesoftware.financisto.http.HttpClientWrapper;
import ru.orangesoftware.financisto.model.Currency;
import ru.orangesoftware.financisto.utils.Logger;

/**
 * Created by vteremasov on 11/8/17.
 */

public class FreeCurrencyRateDownloader extends AbstractMultipleRatesDownloader {

    private final Logger logger = new DependenciesHolder().getLogger();

    private final HttpClientWrapper client;
    private final long dateTime;

    public FreeCurrencyRateDownloader(HttpClientWrapper client, long dateTime) {
        this.client = client;
        this.dateTime = dateTime;
    }

    @Override
    public ExchangeRate getRate(Currency fromCurrency, Currency toCurrency) {
        ExchangeRate rate = createRate(fromCurrency, toCurrency);
        try {
            String s = getResponse(fromCurrency, toCurrency);
            rate.rate = Double.parseDouble(s);
            return rate;
        } catch (Exception e) {
            rate.error = "Unable to get exchange rates: "+e.getMessage();
        }
        return rate;
    }

    @Override
    public ExchangeRate getRate(Currency fromCurrency, Currency toCurrency, long atTime) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    private String getResponse(Currency fromCurrency, Currency toCurrency) throws Exception {
        String url = buildUrl(fromCurrency, toCurrency);
        logger.i(url);
        JSONObject jsonObject = client.getAsJson(url);
        logger.i(jsonObject.getString(toCurrency.name));
        return jsonObject.getString(toCurrency.name);
    }

    private ExchangeRate createRate(Currency fromCurrency, Currency toCurrency) {
        ExchangeRate rate = new ExchangeRate();
        rate.fromCurrencyId = fromCurrency.id;
        rate.toCurrencyId = toCurrency.id;
        rate.date = dateTime;
        return rate;
    }

    private String buildUrl (Currency fromCurrency, Currency toCurrency) {
        return "https://freecurrencyrates.com/api/action.php?s=fcr&iso="+toCurrency.name+"&f="+fromCurrency.name+"&v=1&do=cvals";
    }
}

File: app/src/main/java/ru/orangesoftware/financisto/rates/FreeCurrencyRateDownloader.kt
package ru.orangesoftware.financisto.rates

import kotlin.time.Clock
import ru.orangesoftware.financisto.http.HttpClientWrapper
import ru.orangesoftware.financisto.model.Currency
import ru.orangesoftware.financisto.utils.Logger

class FreeCurrencyRateDownloader(
    private val httpClientWrapper: HttpClientWrapper,
    private val logger: Logger,
    private val clock: Clock,
) : BaseExchangeRateDownloader(clock) {

    override fun getRate(fromCurrency: Currency, toCurrency: Currency): ExchangeRate {
        val exchangeRate = createRate(fromCurrency, toCurrency)
        exchangeRate.safeExecute {
            val s = getResponse(fromCurrency, toCurrency)
            exchangeRate.rate = s.toDouble()
        }
        return exchangeRate
    }

    override fun getRate(fromCurrency: Currency, toCurrency: Currency, atTime: Long): ExchangeRate {
        throw UnsupportedOperationException("Historical rates not supported yet by this downloader")
    }

    private fun getResponse(fromCurrency: Currency, toCurrency: Currency): String {
        val url = buildUrl(fromCurrency, toCurrency)
        logger.i("Downloading rates from FreeCurrencyRates: $url")
        val jsonObject = httpClientWrapper.getAsJson(url)
        val result = jsonObject.getString(toCurrency.name)
        logger.i("Response: $result")
        return result
    }

    private fun buildUrl(fromCurrency: Currency, toCurrency: Currency): String {
        return "https://freecurrencyrates.com/api/action.php?s=fcr&iso=${toCurrency.name}&f=${fromCurrency.name}&v=1&do=cvals"
    }
}